src/Client.php

```diff
@@ -423,40 +424,55 @@ private function sharedContextFactory($initializer, $argv = null)
     }
 
     /**
-     * Creates a new pipeline context and returns it, or returns the results of
-     * a pipeline executed inside the optionally provided callable object.
+     * Creates a new pipeline context.
      *
-     * @param mixed ... Array of options, a callable for execution, or both.
+     * This method returns the pipeline object unless a callable is provided by
+     * the user. In this case the user-supplied callable receives the pipeline
+     * instance as the only argument and the pipeline will be executed by the
+     * client at the end, returning the responses read back from the server.
+     *
+     * @param mixed ... Array of options, a callable for execution, or both
      *
-     * @return Pipeline|array
+     * @return Pipeline|Traversable|array
      */
     public function pipeline(/* arguments */)
     {
         return $this->sharedContextFactory('createPipeline', func_get_args());
     }
 
     /**
-     * Actual pipeline context initializer method.
+     * Pipeline context initializer method.
      *
-     * @param array $options  Options for the context.
-     * @param mixed $callable Optional callable used to execute the context.
+     * @param array    $options  Optional configuration for the context
+     * @param callable $callable Optional callable used to execute the context
      *
-     * @return Pipeline|array
+     * @return Pipeline|Traversable|array
      */
-    protected function createPipeline(array $options = null, $callable = null)
+    protected function createPipeline(array $options = null, callable $callable = null)
     {
-        if (isset($options['atomic']) && $options['atomic']) {
-            $class = 'Predis\Pipeline\Atomic';
-        } elseif (isset($options['fire-and-forget']) && $options['fire-and-forget']) {
-            $class = 'Predis\Pipeline\FireAndForget';
+        if (isset($options['traversable']) && $options['traversable']) {
+            $traversable = true;
         } else {
-            $class = 'Predis\Pipeline\Pipeline';
+            $traversable = false;
         }
 
-        /*
-         * @var ClientContextInterface
-         */
-        $pipeline = new $class($this);
+        if (isset($options['fire-and-forget']) && $options['fire-and-forget']) {
+            $queue = new Queue\FireAndForget();
+        } else {
+            $queue = new Queue\Basic();
+        }
+
+        if (isset($options['atomic']) && $options['atomic']) {
+            $queue = new Queue\Atomic($queue);
+        }
```

Comment:
We could apply the alternative suggestion from above here:

```php
$queue = new Queue\Basic();

if (isset($options['fire-and-forget']) && $options['fire-and-forget']) {
    $queue = new Queue\FireAndForget();
}

if (isset($options['atomic']) && $options['atomic']) {
    $queue = new Queue\Atomic($queue);
}
```

I believe this will make it easier to read as well.

Reply:
In this case I'd prefer to avoid instantiating `Queue\Basic` when not needed. Maybe we can use the null coalesce operator here, too, to get something like this:

```php
// This line on top along with $traversable = ...
$fireandforget = $options['fire-and-forget'] ?? false;

// This line just before checking for atomic.
$queue = $fireandforget ? new Queue\FireAndForget() : new Queue\Basic();
```

How about it?

Reply:
Yes, makes perfect sense

Reply:
So in the end I guess we can rewrite `Predis\Client::createPipeline()` like this:

```php
protected function createPipeline(array $options = null, callable $callable = null)
{
    $traversable = (bool) ($options['traversable'] ?? false);
    $isFireAndForget = (bool) ($options['fire-and-forget'] ?? false);
    $isAtomic = (bool) ($options['atomic'] ?? false);
    $throwExceptions = (bool) ($options['exceptions'] ?? $this->getOptions()->exceptions);

    $queue = $isFireAndForget ? new Queue\FireAndForget() : new Queue\Basic();

    if ($isAtomic) {
        $queue = new Queue\Atomic($queue);
    }

    $pipeline = new Pipeline($this, $queue, $traversable);
    $pipeline->setThrowOnErrorResponse($throwExceptions);

    if (isset($callable)) {
        return $pipeline->execute($callable);
    }

    return $pipeline;
}
```

Null coalesce operator galore! :rocket:

Reply:
HAHAHA!

Reply:
Fixed.